In [140]:
import pandas as pd
import numpy as np
h_data = pd.read_csv("housing.csv")
h_data.head()

id             date   price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900         3       1.00         1180   
1  6414100192  20141209T000000  538000         3       2.25         2570   
2  5631500400  20150225T000000  180000         2       1.00          770   
3  2487200875  20141209T000000  604000         4       3.00         1960   
4  1954400510  20150218T000000  510000         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

# Description of the Dataset:
    This dataset contains these columns: id, date, price, bedrooms, bathrooms, sqft_living,
    sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built,
    yr_renovated, zip code, lat, long, sqft_living15, sqft_lot15. 

# 1. How many houses have a waterfront?
    A. 140
    B. 340
    C. 323
    D. 163
    2. Which zip code

In [8]:
filter_waterfront = h_data["waterfront"]>0
h_data[filter_waterfront]["waterfront"].count()

163

# Answer D

# 2. Which zip code has the costliest house?
    A. 98456
    B. 98102
    C. 98283
    D. 89034

In [21]:
filter_zipcode = h_data.groupby(["zipcode"])["price"].max()
filter_zipcode = filter_zipcode.to_frame(name = "max_price").reset_index()
filter_zipcode_max_price =filter_zipcode["max_price"] ==filter_zipcode["max_price"].max() 
filter_zipcode[filter_zipcode_max_price]

zipcode  max_price
41    98102    7700000

# Answer B

# 3. How many houses are having grade 10?
    A. 1133
    B. 1134
    C. 1135
    D. 1132

In [23]:
filter_grade_10 = h_data["grade"]==10
h_data[filter_grade_10].count()["id"]

1134

# Answer B

In [350]:
pd.DataFrame(sample_tuple, columns=['Number', 'Number_text', 'txtlen'])

Number Number_text  txtlen
0       1         one       3
1       2         two       3
2       3       Three       5
3       4        Four       4
4       5        Five       4

# 4. How many null values are there in the dataset?
    A. 20
    B. 19
    C. 0
    D. 4

In [28]:
h_data.isna().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

# Answer c

# 5. Does this 9126100861 customer have a waterfront?
    A. Yes
    B. No

In [44]:
filter_customer = ((h_data["id"]==9126100861) & (h_data["waterfront"] > 0))
h_data[filter_customer]
# h_data

Empty DataFrame
Columns: [id, date, price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15]
Index: []

[0 rows x 21 columns]

# Answer B

# 6. How many houses have 3 views?
    A. 340
    B. 550
    C. 510
    D. 90

In [47]:
filter_view = h_data["view"] ==3 
h_data[filter_view].count()["id"]

510

# Answer C

# 7. What is the lowest price of the house?
    A. 45000
    B. 28000
    C. 75000
    D. 25000

In [50]:
filter_zipcode = h_data.min()
filter_zipcode["price"]

75000

# Answer C

# 8. Which zip code area has the cheapest house?
    A. 98000
    B. 98022
    C. 80000
    D. 94500

In [51]:
filter_zipcode = h_data.groupby(["zipcode"])["price"].min()
filter_zipcode = filter_zipcode.to_frame(name = "min_price").reset_index()
filter_zipcode_min_price =filter_zipcode["min_price"] ==filter_zipcode["min_price"].min() 
filter_zipcode[filter_zipcode_min_price]

zipcode  min_price
12    98022      75000

# Answer B

# 9. Which area has the biggest sqft_living?
    A. 98053
    B. 98055
    C. 98035
    D. None of the above

In [56]:
filter_zipcode = h_data.groupby(["zipcode"])["sqft_living"].max()
filter_zipcode = filter_zipcode.to_frame(name = "max_sqft_living").reset_index()
filter_zipcode_max_msqft_living =filter_zipcode["max_sqft_living"] ==filter_zipcode["max_sqft_living"].max() 
filter_zipcode[filter_zipcode_max_msqft_living]

zipcode  max_sqft_living
29    98053            13540

# Answer A

# 10. In which year the costliest house has been built?
    A. 2000
    B. 1992
    C. 1911
    D. 1910

In [102]:
filter_yr_built = h_data.groupby(["yr_built"])["long"].sum()
filter_yr_built = filter_yr_built.to_frame(name = "max_price").reset_index()
filter_yr_built.max()

yr_built     2015.000
max_price   -2566.857
dtype: float64

# Answer No of them

In [100]:
h_data

id             date   price  bedrooms  bathrooms  sqft_living  \
0      7129300520  20141013T000000  221900         3       1.00         1180   
1      6414100192  20141209T000000  538000         3       2.25         2570   
2      5631500400  20150225T000000  180000         2       1.00          770   
3      2487200875  20141209T000000  604000         4       3.00         1960   
4      1954400510  20150218T000000  510000         3       2.00         1680   
...           ...              ...     ...       ...        ...          ...   
21608   263000018  20140521T000000  360000         3       2.50         1530   
21609  6600060120  20150223T000000  400000         4       2.50         2310   
21610  1523300141  20140623T000000  402101         2       0.75         1020   
21611   291310100  20150116T000000  400000         3       2.50         1600   
21612  1523300157  20141015T000000  325000         2       0.75         1020   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
0          5650     1.0           0     0  ...      7        1180   
1          7242     2.0           0     0  ...      7        2170   
2         10000     1.0           0     0  ...      6         770   
3          5000     1.0           0     0  ...      7        1050   
4          8080     1.0           0     0  ...      8        1680   
...         ...     ...         ...   ...  ...    ...         ...   
21608      1131     3.0           0     0  ...      8        1530   
21609      5813     2.0           0     0  ...      8        2310   
21610      1350     2.0           0     0  ...      7        1020   
21611      2388     2.0           0     0  ...      8        1600   
21612      1076     2.0           0     0  ...      7        1020   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0                  0      1955             0    98178  47.5112 -122.257   
1                400      1951          1991    98125  47.7210 -122.319   
2                  0      1933             0    98028  47.7379 -122.233   
3                910      1965             0    98136  47.5208 -122.393   
4                  0      1987             0    98074  47.6168 -122.045   
...              ...       ...           ...      ...      ...      ...   
21608              0      2009             0    98103  47.6993 -122.346   
21609              0      2014             0    98146  47.5107 -122.362   
21610              0      2009             0    98144  47.5944 -122.299   
21611              0      2004             0    98027  47.5345 -122.069   
21612              0      2008             0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
0               1340        5650  
1               1690        7639  
2               2720        8062  
3               1360        5000  
4               1800        7503  
...              ...         ...  
21608           1530        1509  
21609           1830        7200  
21610           1020        2007  
21611           1410        1287  
21612           1020        1357  

[21613 rows x 21 columns]

# 11. Which of the following data types can a Pandas Series have?
    A. Int
    B. Str
    C. Float
    D. All of the above

# Answer D

# 12. Which of the following is having a one-dimensional array?
    A. Series data type
    B. DataFrame data type
    C. Both of the above
    D. None of the above

# Answer A

# 13. What value will you get on the left side after printing a series format data?
    A. Data
    B. Value
    C. Index
    D. All of the above

# Answer C

# 14. Which of the following is considered as data in pandas?
    A. Dictionary
    B. Boolean
    C. Nd-array
    D. All of the above

# Answer D

# 15. In what format the keys will get converted into when we convert the dictionary data into data frame format?
    A. Rows
    B. Columns
    C. Indexes
    D. Records

# Answer B

# 16. Write a function to perform the following operation:
    Increase the grade of the house by 1 if the sqft of the house is greater than 0 and less than
    equal to 400. If the total sqft of the house is greater than 400, increase the grade of the house
    by 2.
    After creating the above function, which syntax will be using when using apply function to run
    the above created function in the dataset?
        A. data.apply(your_function_name(arg1))
        B. data.apply_fun(function_name())
        C. data.app(data)
        D. None of the above

In [151]:
import pandas as pd
import numpy as np
h_data = pd.read_csv("housing.csv")

def increase_grade(grade, sqft):
    if (sqft > 0) and (sqft <=400):
        return grade + 1
    else:
        return grade + 2
    
h_data["grade"] = h_data.apply(lambda x: increase_grade(x["grade"], x["sqft_living"]), axis = 1)
h_data.head()

id             date   price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900         3       1.00         1180   
1  6414100192  20141209T000000  538000         3       2.25         2570   
2  5631500400  20150225T000000  180000         2       1.00          770   
3  2487200875  20141209T000000  604000         4       3.00         1960   
4  1954400510  20150218T000000  510000         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      9        1180              0   
1      7242     2.0           0     0  ...      9        2170            400   
2     10000     1.0           0     0  ...      8         770              0   
3      5000     1.0           0     0  ...      9        1050            910   
4      8080     1.0           0     0  ...     10        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

# Answer A

# 17. How to check the duplicate values on id,grade and location?
    A. df.duplicated(subset=['id','zipcode','grade'])
    B. df.duplicate(subset=['id','zipcode','grade'])
    C. df[df.duplicated(['id','zipcode','grade’])]
    D. All of the above

In [160]:
import pandas as pd
import numpy as np
df = pd.read_csv("housing.csv")
df[df.duplicated(['id','zipcode','grade'])]


id             date    price  bedrooms  bathrooms  sqft_living  \
94     6021501535  20141223T000000   700000         3       1.50         1580   
314    4139480200  20141209T000000  1400000         4       3.25         4290   
325    7520000520  20150311T000000   240500         2       1.00         1240   
346    3969300030  20141229T000000   239900         4       1.00         1000   
372    2231500030  20150324T000000   530000         4       2.25         2180   
...           ...              ...      ...       ...        ...          ...   
20181  7853400250  20150219T000000   645000         4       3.50         2910   
20613  2724049222  20141201T000000   220000         2       2.50         1000   
20670  8564860270  20150330T000000   502000         4       2.50         2680   
20780  6300000226  20150504T000000   380000         4       1.00         1200   
21581  7853420110  20150504T000000   625000         3       3.00         2780   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
94         5000     1.0           0     0  ...      8        1290   
314       12103     1.0           0     3  ...     11        2690   
325       12092     1.0           0     0  ...      6         960   
346        7134     1.0           0     0  ...      6        1000   
372       10754     1.0           0     0  ...      7        1100   
...         ...     ...         ...   ...  ...    ...         ...   
20181      5260     2.0           0     0  ...      9        2910   
20613      1092     2.0           0     0  ...      7         990   
20670      5539     2.0           0     0  ...      8        2680   
20780      2171     1.5           0     0  ...      7        1200   
21581      6000     2.0           0     0  ...      9        2780   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
94               290      1939             0    98117  47.6870 -122.386   
314             1600      1997             0    98006  47.5503 -122.102   
325              280      1922          1984    98146  47.4957 -122.352   
346                0      1943             0    98178  47.4897 -122.240   
372             1080      1954             0    98133  47.7711 -122.341   
...              ...       ...           ...      ...      ...      ...   
20181              0      2012             0    98065  47.5168 -121.883   
20613             10      2004             0    98118  47.5419 -122.271   
20670              0      2013             0    98045  47.4759 -121.734   
20780              0      1933             0    98133  47.7076 -122.342   
21581              0      2013             0    98065  47.5184 -121.886   

       sqft_living15  sqft_lot15  
94              1570        4500  
314             3860       11244  
325             1820        7460  
346             1020        7138  
372             1810        6929  
...              ...         ...  
20181           2910        5260  
20613           1330        1466  
20670           2680        5992  
20780           1130        1598  
21581           2850        6000  

[177 rows x 21 columns]

# Answer C

# 18. Which of the following code will help to display the 3rd, 4th and 5th rows from the 6th to 9th columns of data frame data?
    A. data.loc[3:6, 6:10]
    B. data.iloc[3;6,6;10]
    C. data.iloc[3:6,6:10]
    D. None of the above

In [193]:
df.head()

id             date   price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900         3       1.00         1180   
1  6414100192  20141209T000000  538000         3       2.25         2570   
2  5631500400  20150225T000000  180000         2       1.00          770   
3  2487200875  20141209T000000  604000         4       3.00         1960   
4  1954400510  20150218T000000  510000         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [196]:
df.iloc[3:6,6:10]

sqft_lot  floors  waterfront  view
3      5000     1.0           0     0
4      8080     1.0           0     0
5    101930     1.0           0     0

# Answer C

# 19. Which of the following syntax will display the last two records of df?
    import pandas as pd
    df = pd.DataFrame({'A':[34, 78, 54], 'B':[12, 67, 43]}, index=['r1', 'r2', 'r3'])
    A. df.iloc[:'r3']
    B. df.loc['r2':'r3']
    C. df.iloc['r2':'r3']
    D. df.loc[:'r3']

In [205]:
df = pd.DataFrame({'A':[34, 78, 54], 'B':[12, 67, 43]}, index=['r1', 'r2', 'r3'])
df.loc['r2':'r3']

A   B
r2  78  67
r3  54  43

# Answer B

# 20. Which of the following is/are true about loc in pandas:
    A. Add new rows in the data frames
    B. To change the values of a row to a particular value
    C. To extract values from the particular rows
    D. All of the above

# Answer C

# 21. Change the date column in the format (dd/mm/year) using the pandas to_datetime() Function.
    A. pd.to_datetime(data['date'], format='%Y-%m-%d', utc=False, dayfirst=True)
    B. pd.to_datetime(data['date'])
    C. pd.to_datetime(data['dates'], format='%Y-%m-%d', utc=False, dayfirst=True)
    D. pd.to_datetime(data['date'])

In [347]:
import pandas as pd
import numpy as np
data = pd.read_csv("housing.csv")
data.head()


id             date   price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900         3       1.00         1180   
1  6414100192  20141209T000000  538000         3       2.25         2570   
2  5631500400  20150225T000000  180000         2       1.00          770   
3  2487200875  20141209T000000  604000         4       3.00         1960   
4  1954400510  20150218T000000  510000         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [342]:
data['New_date'] = pd.to_datetime(data['date'], format='%Y/%m/%d', utc=False, dayfirst=True)
# data['New_date'] = pd.to_datetime(data['date'])
# pd.to_datetime(data['New_date'] , format='%b-%y')
# data['New_date'] = df["date"].dt.strftime('%Y%m%d').astype(int)
# data.head()
data['New_date']

0       2014-10-13
1       2014-12-09
2       2015-02-25
3       2014-12-09
4       2015-02-18
           ...    
21608   2014-05-21
21609   2015-02-23
21610   2014-06-23
21611   2015-01-16
21612   2014-10-15
Name: New_date, Length: 21613, dtype: datetime64[ns]

In [345]:
# from datetime import datetime
# data['New_date'] = data['date'].apply(datetime.strftime, format='%Y%m%d').astype(int)
data['New_date'] = pd.to_datetime(data['date'], format='%Y-%m-%d', utc=False, dayfirst=True) & pd.to_datetime(data['date'])
# from datetime import datetime
data['New_date'] 

TypeError: unsupported operand type(s) for &: 'DatetimeArray' and 'DatetimeArray'

# 22. Create a separate data frame that satisfies the conditions below.
    1. Houses built before 1980
    2. Have more than 2 bedrooms
    3. Have more than 2 floors.
    A. data.loc[(data["yr_built"] > 1980) & (data['floors'] > 2) & (data['bedrooms'] > 2)]
    B. data.loc[(data["yr_built"] < 1980) & (data['floors'] > 2) & (data['bedrooms'] > 2)]
    C. data.loc[(data["yr_built"] | 1980) & (data['floors'] > 2) | (data['bedrooms'] > 2)]
    D. data.loc[(data["yr_built"] < 1980) | (data['floors'] > 2) | (data['bedrooms'] < 2)]

# Answer B

# 23. For a given nested list, convert the same into a dataframe.
    sample_list = [['Carl', 22],
    ['Martha', 25],
    ['Calvin', 12],
    ['Stuart', 15]
    ]
    The resulting dataframe must contain the column names as ‘Name’, and ‘Age’ with the
    respective values from the sample_list.
    A. pd.DataFrame(sample_list)
    B. pd.DataFrame(sample_list, column_names=['Name', 'Age'])
    C. pd.DataFrame(sample_list, columns=['Name', 'Age'])
    D. pd.DataFrame(sample_list, column_name=['Name', 'Age'])

In [278]:
sample_list = [['Carl', 22],
['Martha', 25],
['Calvin', 12],
['Stuart', 15]
]

In [279]:
pd.DataFrame(sample_list, columns=['Name', 'Age'])

Name  Age
0    Carl   22
1  Martha   25
2  Calvin   12
3  Stuart   15

# Answer C

# 24. For a given dictionary, convert the same into a dataframe.
    sample_dict = {'Cristiano': ['Ronaldo','Man U', 801],
    'Lionel': ['Messi','PSG', 758],
    'Luis': ['Suarez','Atletico Madrid', 509],
    'Robert': ['Lewandowski','Bayern Munich', 527],
    'Zlatan': ['Ibrahimovic','AC Milan',553]
    }
    A. df1 = pd.Dataframe(sample_dict)
    df1 = df1.transpose()
    df1.reset_index(inplace = True)
    df1.columns = ['First Name','Last Name', 'Club', 'Goals']
    B. df1 = pd.DataFrame(sample_dict)
    df1 = df1.transpose()
    df1.reset_index(inplace = True)
    df1.columns = ['First Name','Last Name', 'Club', 'Goals']
    C. df1 = pd.DataFrame(sample_dict)
    df1.reset_index(inplace = True)
    df1.columns = ['First Name','Last Name', 'Club', 'Goals']
    D. df1 = pd.DataFrame(sample_dict)
    df1 = df1.transpose()
    df1.columns = ['First Name','Last Name', 'Club', 'Goals']

In [276]:
sample_dict = {'Cristiano': ['Ronaldo','Man U', 801],
'Lionel': ['Messi','PSG', 758],
'Luis': ['Suarez','Atletico Madrid', 509],
'Robert': ['Lewandowski','Bayern Munich', 527],
'Zlatan': ['Ibrahimovic','AC Milan',553]
}

df1 = pd.DataFrame(sample_dict)
df1 = df1.transpose()
df1.reset_index(inplace = True)
df1.columns = ['First Name','Last Name', 'Club', 'Goals']
df1

First Name    Last Name             Club Goals
0  Cristiano      Ronaldo            Man U   801
1     Lionel        Messi              PSG   758
2       Luis       Suarez  Atletico Madrid   509
3     Robert  Lewandowski    Bayern Munich   527
4     Zlatan  Ibrahimovic         AC Milan   553

# Answer B

In [351]:
pd.DataFrame(sample_tuple, columns=['Number', 'Number_text', 'txtlen'])

Number Number_text  txtlen
0       1         one       3
1       2         two       3
2       3       Three       5
3       4        Four       4
4       5        Five       4

In [349]:
df1 = pd.DataFrame(sample_dict) 
df1 = df1.transpose()
df1.reset_index(inplace = True) 
df1.columns = ['First Name','Last Name', 'Club', 'Goals']
df1

First Name    Last Name             Club Goals
0  Cristiano      Ronaldo            Man U   801
1     Lionel        Messi              PSG   758
2       Luis       Suarez  Atletico Madrid   509
3     Robert  Lewandowski    Bayern Munich   527
4     Zlatan  Ibrahimovic         AC Milan   553

# 25. For a given tuple, convert the same into a dataframe.
    sample_tuple = ([1, 'one', 3],
    [2, 'two', 3],
    [3, 'Three', 5],
    [4, 'Four', 4],
    [5, 'Five', 4])
    A. pd.Dataframe(sample_tuple, columns=['Number', 'Number_text', 'txtlen'])
    B. pd.DataFrame(sample_tuple, columns=['Number', 'Number_text', 'txtlen'])
    C. pd.dataframe(sample_tuple, columns=['Number', 'Number_text', 'txtlen'])
    D. pd.DataFrame(sample_tup, columns=['Number', 'Number_text', 'txtlen'])

In [290]:
sample_tuple = ([1, 'one', 3],
[2, 'two', 3],
[3, 'Three', 5],
[4, 'Four', 4],
[5, 'Five', 4])

In [291]:
pd.DataFrame(sample_tuple, columns=['Number', 'Number_text', 'txtlen'])

Number Number_text  txtlen
0       1         one       3
1       2         two       3
2       3       Three       5
3       4        Four       4
4       5        Five       4

# Answer B

# 26. Create a separate dataframe that contains houses ordered in ascending or descending order of the prices of each house.
    A. ascending = housing.sort_values('price', ascending=False)
    descending = housing.sort_values('price', ascending=False)
    B. ascending = housing.sort_values('price', ascending=True)
    descending = housing.sort_values('price', ascending=False)
    C. ascending = housing.sort_values('price', ascending=False)
    descending = housing.sort_values('price', ascending=True)
    D. ascending = housing.sort_values('price', ascending=True)
    descending = housing.sort_values('price', ascending=True)

In [296]:
housing = h_data
housing.head()


# B. 
ascending = housing.sort_values('price', ascending=True)
descending = housing.sort_values('price', ascending=False)

# Answer B

In [352]:
ascending = housing.sort_values('price', ascending=True) 
descending = housing.sort_values('price', ascending=False)

# 27. Calculate the mean and standard deviation of all the numerical values in the dataset. For example - the mean for the bedrooms column is 3.370 and the standard deviation is 0.930.
    A. housing.describe()
    B. housing.info()
    C. housing.corr()
    D. housing.std()

In [300]:
housing.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

In [299]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  int64  
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [298]:
housing.corr()

C:\Users\27746\AppData\Local\Temp\ipykernel_7772\1598269145.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  housing.corr()


id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
id             1.000000 -0.016762  0.001286   0.005160    -0.012258 -0.132109   
price         -0.016762  1.000000  0.308350   0.525138     0.702035  0.089661   
bedrooms       0.001286  0.308350  1.000000   0.515884     0.576671  0.031703   
bathrooms      0.005160  0.525138  0.515884   1.000000     0.754665  0.087740   
sqft_living   -0.012258  0.702035  0.576671   0.754665     1.000000  0.172826   
sqft_lot      -0.132109  0.089661  0.031703   0.087740     0.172826  1.000000   
floors         0.018525  0.256794  0.175429   0.500653     0.353949 -0.005201   
waterfront    -0.002721  0.266369 -0.006582   0.063744     0.103818  0.021604   
view           0.011592  0.397293  0.079532   0.187737     0.284611  0.074710   
condition     -0.023783  0.036362  0.028472  -0.124982    -0.058753 -0.008958   
grade          0.008130  0.667434  0.356967   0.664983     0.762704  0.113621   
sqft_above    -0.010842  0.605567  0.477600   0.685342     0.876597  0.183512   
sqft_basement -0.005151  0.323816  0.303093   0.283770     0.435043  0.015286   
yr_built       0.021380  0.054012  0.154178   0.506019     0.318049  0.053080   
yr_renovated  -0.016907  0.126434  0.018841   0.050739     0.055363  0.007644   
zipcode       -0.008224 -0.053203 -0.152668  -0.203866    -0.199430 -0.129574   
lat           -0.001891  0.307003 -0.008931   0.024573     0.052529 -0.085683   
long           0.020799  0.021626  0.129473   0.223042     0.240223  0.229521   
sqft_living15 -0.002901  0.585379  0.391638   0.568634     0.756420  0.144608   
sqft_lot15    -0.138798  0.082447  0.029244   0.087175     0.183286  0.718557   

                 floors  waterfront      view  condition     grade  \
id             0.018525   -0.002721  0.011592  -0.023783  0.008130   
price          0.256794    0.266369  0.397293   0.036362  0.667434   
bedrooms       0.175429   -0.006582  0.079532   0.028472  0.356967   
bathrooms      0.500653    0.063744  0.187737  -0.124982  0.664983   
sqft_living    0.353949    0.103818  0.284611  -0.058753  0.762704   
sqft_lot      -0.005201    0.021604  0.074710  -0.008958  0.113621   
floors         1.000000    0.023698  0.029444  -0.263768  0.458183   
waterfront     0.023698    1.000000  0.401857   0.016653  0.082775   
view           0.029444    0.401857  1.000000   0.045990  0.251321   
condition     -0.263768    0.016653  0.045990   1.000000 -0.144674   
grade          0.458183    0.082775  0.251321  -0.144674  1.000000   
sqft_above     0.523885    0.072075  0.167649  -0.158214  0.755923   
sqft_basement -0.245705    0.080588  0.276947   0.174105  0.168392   
yr_built       0.489319   -0.026161 -0.053440  -0.361417  0.446963   
yr_renovated   0.006338    0.092885  0.103917  -0.060618  0.014414   
zipcode       -0.059121    0.030285  0.084827   0.003026 -0.184862   
lat            0.049614   -0.014274  0.006157  -0.014941  0.114084   
long           0.125419   -0.041910 -0.078400  -0.106500  0.198372   
sqft_living15  0.279885    0.086463  0.280439  -0.092824  0.713202   
sqft_lot15    -0.011269    0.030703  0.072575  -0.003406  0.119248   

               sqft_above  sqft_basement  yr_built  yr_renovated   zipcode  \
id              -0.010842      -0.005151  0.021380     -0.016907 -0.008224   
price            0.605567       0.323816  0.054012      0.126434 -0.053203   
bedrooms         0.477600       0.303093  0.154178      0.018841 -0.152668   
bathrooms        0.685342       0.283770  0.506019      0.050739 -0.203866   
sqft_living      0.876597       0.435043  0.318049      0.055363 -0.199430   
sqft_lot         0.183512       0.015286  0.053080      0.007644 -0.129574   
floors           0.523885      -0.245705  0.489319      0.006338 -0.059121   
waterfront       0.072075       0.080588 -0.026161      0.092885  0.030285   
view             0.167649       0.276947 -0.053440      0.103917  0.084827   
condition       -0.158214       0.174105 -0.361417     -0.060618  0.

In [297]:
housing.std()

C:\Users\27746\AppData\Local\Temp\ipykernel_7772\2264156241.py:1: FutureWarning: The default value of numeric_only in DataFrame.std is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  housing.std()


id               2.876566e+09
price            3.671272e+05
bedrooms         9.300618e-01
bathrooms        7.701632e-01
sqft_living      9.184409e+02
sqft_lot         4.142051e+04
floors           5.399889e-01
waterfront       8.651720e-02
view             7.663176e-01
condition        6.507430e-01
grade            1.175459e+00
sqft_above       8.280910e+02
sqft_basement    4.425750e+02
yr_built         2.937341e+01
yr_renovated     4.016792e+02
zipcode          5.350503e+01
lat              1.385637e-01
long             1.408283e-01
sqft_living15    6.853913e+02
sqft_lot15       2.730418e+04
dtype: float64

# Answer A

# 28. Perform the following operations on the pandas dataframe.
    1. Get the data starting from the rows 25 to 35.
    2. Get the price of the houses located at the longitude--122.045 latitude- 47.6168.
    A. data[25:35], data.loc[(housing["lat"] == 47.6168) | (data["long"] == -122.045)]
    B. data[25:35], data.loc[(housing["lat"] == 48.6168) & (data["long"] == -122.045)]
    C. data[25:36], data.loc[(housing["lat"] == 47.6168) & (data["long"] == -122.045)]
    D. data[24:36], data.loc[(housing["lat"] == 47.6168) & (data["long"] == -122.045)]

In [308]:
# A. 
# data[25:35], data.loc[(housing["lat"] == 47.6168) | (data["long"] == -122.045)]
# B. 
# data[25:35], data.loc[(housing["lat"] == 48.6168) & (data["long"] == -122.045)]
# C. 
data[25:36], data.loc[(housing["lat"] == 47.6168) & (data["long"] == -122.045)]
# D. 
# data[24:36], data.loc[(housing["lat"] == 47.6168) & (data["long"] == -122.045)]

(            id             date   price  bedrooms  bathrooms  sqft_living  \
 25  1202000200  20141103T000000  233000         3       2.00         1710   
 26  1794500383  20140626T000000  937000         3       1.75         2450   
 27  3303700376  20141201T000000  667000         3       1.00         1400   
 28  5101402488  20140624T000000  438000         3       1.75         1520   
 29  1873100390  20150302T000000  719000         4       2.50         2570   
 30  8562750320  20141110T000000  580500         3       2.50         2320   
 31  2426039314  20141201T000000  280000         2       1.50         1190   
 32   461000390  20140624T000000  687500         4       1.75         2330   
 33  7589200193  20141110T000000  535000         3       1.00         1090   
 34  7955080270  20141203T000000  322500         4       2.75         2060   
 35  9547205180  20140613T000000  696000         3       2.50         2300   
 
     sqft_lot  floors  waterfront  view  ...  sqft_above  sqft

# Answer C

# 29. Create a new column with the floor area(sqft_living, sqft_lot, sqft_above, sqft_basement, all combined in one column).
    A. data['Floor Area'] = data['sqft_living'], data['sqft_lot'], data['sqft_basement'],
    data['sqft_above']
    B. df['Floor Area'] = data['sqft_living'] + data['sqft_lot'] + data['sqft_basement'] +
    data['sqft_above']
    C. data['Floor Area'] = data['sqft_living', 'sqft_lot','sqft_basement','sqft_above']
    D. data['Floor Area'] = data['sqft_liv', 'sqft_lot','sqft_base','sqft_above']

In [317]:
df['Floor Area'] = data['sqft_living'] + data['sqft_lot'] + data['sqft_basement'] + data['sqft_above']
df

DOB  Floor Area
0  26/1/2016        8010
1  26/1/2016       12382

# Answer B

# 30. Perform the following operations on the dataframes given below.
    A = pd.DataFrame([['Carl', 22],['Martha', 25],['Calvin', 12],['Stuart', 15]], columns=[‘Name’, ‘Age’])
    B = pd.DataFrame([['Melvin', 25],['Martha', 34],['Lewis', 32],['Leo', 25]], columns=[‘Name’, ‘Age’])
    1. Left Outer Join
    2. Outer Join
    3. Inner Join
    4. Right Outer Join
    A. inner = pd.merge(A, B, on='Age', how='inner_join')
    outer = pd.merge(A, B, on='Age', how='outer_join')
    left_outer = pd.merge(A, B, on='Age', how='left_join')
    right_outer = pd.merge(A, B, on='Age', how='right_join')
    B. inner = pd.merge(A, B, on='Age', how='inner')
    outer = pd.merge(A, B, on='Age', how='outer')
    left_outer = pd.merge(A, B, on='Age', how='left_outer')
    right_outer = pd.merge(A, B, on='Age', how='right_outer')
    C. inner = pd.merge(A, B, on='Age', how='inner')
    outer = pd.merge(A, B, on='Age', how='outer')
    left_outer = pd.merge(A, B, on='Age', how='left')
    right_outer = pd.merge(A, B, on='Age', how='right')
    D. inner = pd.merge(A, B, on='Age', how='inner')
    outer = pd.merge(A, B, on='Age', how='outer')
    left_outer = pd.merge(A, B, on='Age', how='outer_left')
    right_outer = pd.merge(A, B, on='Age', how='outer_right')

In [323]:
A = pd.DataFrame([['Carl', 22],['Martha', 25],['Calvin', 12],['Stuart', 15]], columns=['Name', "Age"])
B = pd.DataFrame([['Melvin', 25],['Martha', 34],['Lewis', 32],['Leo', 25]], columns=['Name', "Age"])
A

Name  Age
0    Carl   22
1  Martha   25
2  Calvin   12
3  Stuart   15

In [324]:
B

Name  Age
0  Melvin   25
1  Martha   34
2   Lewis   32
3     Leo   25

In [335]:
inner = pd.merge(A, B, on='Age', how='inner')
outer = pd.merge(A, B, on='Age', how='outer')
left_outer = pd.merge(A, B, on='Age', how='left')
right_outer = pd.merge(A, B, on='Age', how='right')
right_outer

Name_x  Age  Name_y
0  Martha   25  Melvin
1     NaN   34  Martha
2     NaN   32   Lewis
3  Martha   25     Leo

# Answer C

In [ ]:
C. inner = pd.merge(A, B, on='Age', how='inner')
outer = pd.merge(A, B, on='Age', how='outer')
left_outer = pd.merge(A, B, on='Age', how='left')
right_outer = pd.merge(A, B, on='Age', how='right')

inner = pd.merge(A, B, on='Age', how='inner') 
outer = pd.merge(A, B, on='Age', how='outer') 
left_outer = pd.merge(A, B, on='Age', how='left') 
right_outer = pd.merge(A, B, on='Age', how='right')

# 31. What will be the correlation between the columns sqft_living and sqft_above?
    A. 0.702035
    B. 0.754665
    C. 0.876597
    D. 0.303093

In [336]:
housing.corr()

C:\Users\27746\AppData\Local\Temp\ipykernel_7772\1598269145.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  housing.corr()


id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
id             1.000000 -0.016762  0.001286   0.005160    -0.012258 -0.132109   
price         -0.016762  1.000000  0.308350   0.525138     0.702035  0.089661   
bedrooms       0.001286  0.308350  1.000000   0.515884     0.576671  0.031703   
bathrooms      0.005160  0.525138  0.515884   1.000000     0.754665  0.087740   
sqft_living   -0.012258  0.702035  0.576671   0.754665     1.000000  0.172826   
sqft_lot      -0.132109  0.089661  0.031703   0.087740     0.172826  1.000000   
floors         0.018525  0.256794  0.175429   0.500653     0.353949 -0.005201   
waterfront    -0.002721  0.266369 -0.006582   0.063744     0.103818  0.021604   
view           0.011592  0.397293  0.079532   0.187737     0.284611  0.074710   
condition     -0.023783  0.036362  0.028472  -0.124982    -0.058753 -0.008958   
grade          0.008130  0.667434  0.356967   0.664983     0.762704  0.113621   
sqft_above    -0.010842  0.605567  0.477600   0.685342     0.876597  0.183512   
sqft_basement -0.005151  0.323816  0.303093   0.283770     0.435043  0.015286   
yr_built       0.021380  0.054012  0.154178   0.506019     0.318049  0.053080   
yr_renovated  -0.016907  0.126434  0.018841   0.050739     0.055363  0.007644   
zipcode       -0.008224 -0.053203 -0.152668  -0.203866    -0.199430 -0.129574   
lat           -0.001891  0.307003 -0.008931   0.024573     0.052529 -0.085683   
long           0.020799  0.021626  0.129473   0.223042     0.240223  0.229521   
sqft_living15 -0.002901  0.585379  0.391638   0.568634     0.756420  0.144608   
sqft_lot15    -0.138798  0.082447  0.029244   0.087175     0.183286  0.718557   

                 floors  waterfront      view  condition     grade  \
id             0.018525   -0.002721  0.011592  -0.023783  0.008130   
price          0.256794    0.266369  0.397293   0.036362  0.667434   
bedrooms       0.175429   -0.006582  0.079532   0.028472  0.356967   
bathrooms      0.500653    0.063744  0.187737  -0.124982  0.664983   
sqft_living    0.353949    0.103818  0.284611  -0.058753  0.762704   
sqft_lot      -0.005201    0.021604  0.074710  -0.008958  0.113621   
floors         1.000000    0.023698  0.029444  -0.263768  0.458183   
waterfront     0.023698    1.000000  0.401857   0.016653  0.082775   
view           0.029444    0.401857  1.000000   0.045990  0.251321   
condition     -0.263768    0.016653  0.045990   1.000000 -0.144674   
grade          0.458183    0.082775  0.251321  -0.144674  1.000000   
sqft_above     0.523885    0.072075  0.167649  -0.158214  0.755923   
sqft_basement -0.245705    0.080588  0.276947   0.174105  0.168392   
yr_built       0.489319   -0.026161 -0.053440  -0.361417  0.446963   
yr_renovated   0.006338    0.092885  0.103917  -0.060618  0.014414   
zipcode       -0.059121    0.030285  0.084827   0.003026 -0.184862   
lat            0.049614   -0.014274  0.006157  -0.014941  0.114084   
long           0.125419   -0.041910 -0.078400  -0.106500  0.198372   
sqft_living15  0.279885    0.086463  0.280439  -0.092824  0.713202   
sqft_lot15    -0.011269    0.030703  0.072575  -0.003406  0.119248   

               sqft_above  sqft_basement  yr_built  yr_renovated   zipcode  \
id              -0.010842      -0.005151  0.021380     -0.016907 -0.008224   
price            0.605567       0.323816  0.054012      0.126434 -0.053203   
bedrooms         0.477600       0.303093  0.154178      0.018841 -0.152668   
bathrooms        0.685342       0.283770  0.506019      0.050739 -0.203866   
sqft_living      0.876597       0.435043  0.318049      0.055363 -0.199430   
sqft_lot         0.183512       0.015286  0.053080      0.007644 -0.129574   
floors           0.523885      -0.245705  0.489319      0.006338 -0.059121   
waterfront       0.072075       0.080588 -0.026161      0.092885  0.030285   
view             0.167649       0.276947 -0.053440      0.103917  0.084827   
condition       -0.158214       0.174105 -0.361417     -0.060618  0.

# Answer C

# 32. If the correlation between the columns bathrooms and sqft_living is 0.754665, what all interpretations can be made about the two columns?
    A. A positive correlation between the two columns
    B. The columns show perfect correlation.
    C. A strong negative correlation between the two columns
    D. No correlation between the two columns.

# Answer A

The correlation coefficient between the columns bathrooms and sqft_living is 0.7546651. This indicates that there is a strong positive correlation between the two columns. In other words, as the number of bathrooms increases, so does the square footage of living space. However, it is important to note that correlation does not imply causation. There may be other factors that influence both the number of bathrooms and the square footage of living space. For example, the number of bedrooms or the location of the house may also be important factors to consider when analyzing house prices.